app/src/main/java/ch/epfl/sweng/erpa/operations/LoggedUserCoordinator.java

```diff
@@ -66,49 +65,34 @@ public void unsetCurrentLoggedUser() {
+            profile.setUuid(userUuid);
+            String base64PublicKey = Base64.encodeToString(CryptoGrenouille
+                .ed25519KeyPairFromPassphraseAndUserUuid(password, userUuid)
+                .getPublic().getEncoded(), Base64.NO_WRAP);
+            UserAuth userAuth = new UserAuth(userUuid, base64PublicKey, CryptoGrenouille.AUTH_STRATEGY_NAME);
+            ups.registerAuth(userAuth);
+            ups.registerUserProfile(profile);
+            this.tryLogin(ts, username, password, successCallback, errorHandler);
+            return null;
+        }, o -> {
+        }, errorHandler);
     }
 
-    public AsyncTask<Void, Void, Exceptional<String>> tryLogin(AsyncTaskService ts,
-                                                               String usernameText, String passwordText,
-                                                               Runnable successCallback,
-                                                               Consumer<Throwable> errorHandler) {
-        // @formatter:off Think of this like a scala query with the arrows graphically reversed...
-        AsyncTaskService.Runner<String> login = ts.create(() -> ups.getUuidForUsername(usernameText),
-            failIfNull(String.format("Could not retrieve UserUUID for username '%s'", usernameText),
-            userUuid -> ts.run(() -> {
-                // TODO(@Roos): Call the GetSessionToken endpoint with the password info
-                return new UserSessionToken(userUuid, userUuid);
-            },
-            userSessionToken -> ts.run(() -> ups.getUsernameFromUserUuid(userUuid), failIfNotFound(userUuid,
-            (Username username) -> ts.run(() -> ups.getUserProfile(userUuid), failIfNotFound(userUuid,
-            (UserProfile userProfile) -> {
-                this.publishLoggedUser(username, userSessionToken, userProfile);
-                successCallback.run();
-            })))))));
-        // @formatter:on
-        login.setThrowableConsumer(errorHandler);
-        login.execute();
-        return login;
+    public AsyncTask<Void, Void, Exceptional<String>> tryLogin(
+        AsyncTaskService ts, String usernameText, String password,
+        Runnable successCallback, Consumer<Throwable> errorHandler) {
+        return ts.run(() -> {
+            String userUuid = ups.getUuidForUsername(usernameText);
+            String b64AuthChallenge = ups.getBase64AuthenticationChallenge(userUuid);
+            String challengeResponse = new CryptoGrenouille(password, userUuid).signBase64Encoded(b64AuthChallenge);
+            UserSessionToken userSessionToken = ups.getSessionToken(userUuid, challengeResponse);
+            Username username = ups.getUsernameFromUserUuid(userUuid).get();
+            UserProfile userProfile = ups.getUserProfile(userUuid).get();
+            this.publishLoggedUser(username, userSessionToken, userProfile);
+            return null;
+        }, o -> successCallback.run(), errorHandler);
```

Comment:
Why are you staying on the same line here when you have code after the lambda and going to another line on line 80 when the lambda is empty ?

Reply:
Because expanded lambdas must not start and end in the same line. Also, the only way of getting a nop lambda is by using an expanded lambda with  an empty body.